lib/RefBackend/RefBackend.cpp

```diff
@@ -95,28 +95,45 @@ static LogicalResult mungeFunction(
   SmallVector<Operation *> toErase;
   bool hadError = false;
   func.walk([&](ReturnOp op) {
-    auto memRefType = op.getOperandTypes()[0].dyn_cast<MemRefType>();
-    if (!memRefType) {
+    auto returnType = op.getOperandTypes()[0];
+    if (returnType.isa<MemRefType>()) {
+      auto elemType = returnType.dyn_cast<MemRefType>().getElementType();
+      auto unRankedMemrefType = UnrankedMemRefType::get(elemType, 0);
+      auto it = consumeFuncReturnFuncs.find(unRankedMemrefType);
+      if (op.getNumOperands() != 1 || it == consumeFuncReturnFuncs.end()) {
+        hadError = true;
+        op.emitError("must have one return value");
+        return;
+      }
+      b.setInsertionPoint(op);
+      auto cast = b.create<memref::CastOp>(
+          op.getLoc(), op.getOperand(0),
+          getAbiTypeForMemRef(op.getOperandTypes()[0]));
+      b.create<mlir::CallOp>(op.getLoc(),
+                             consumeFuncReturnFuncs[unRankedMemrefType],
+                             cast.getResult());
+      b.create<mlir::ReturnOp>(op.getLoc());
+      toErase.push_back(op);
+
+    } else if (returnType.isa<IntegerType>() || returnType.isa<FloatType>()) {
+      auto it = consumeFuncReturnFuncs.find(returnType);
+      if (op.getNumOperands() != 1 || it == consumeFuncReturnFuncs.end()) {
+        hadError = true;
+        op.emitError("must have one return value");
+        return;
+      }
+
+      b.setInsertionPoint(op);
+      b.create<mlir::CallOp>(op.getLoc(), consumeFuncReturnFuncs[returnType],
+                             op.getOperand(0));
+      b.create<mlir::ReturnOp>(op.getLoc());
+      toErase.push_back(op);
+
+    } else {
```

Comment:
nit: It seems to me that if a type is not valid, then it should not be in the `consumeFuncReturnFuncs` map, so the error checking above makes this error checking not necessary. I would just move the error message from here to the error checking for the `it` variable. Knowing this will also simplify the `if` statements quite a bit because if the return type is in `consumeFuncReturnFuncs` and it is not a complex type (like `memref`), then you can assume that it is a valid simple type, which means the `if` check for `IntegerType` and `FloatType` above is not needed. In other words, the case where you have a simple type should be your `else` branch.

(by "simple type" here I mean types that don't need any extra ops other than `CallOp` and `ReturnOp`)